subscriber/src/daily_operations.rs

```diff
@@ -0,0 +1,227 @@
+                }
+            };
+
+            progress.current_index += 1;
+
+            let subscription_type = self.subscription_type(user_id, service_index).get();
+            let next_action_epoch =
+                self.get_next_action_epoch(user_id, service_index, subscription_type);
+            if next_action_epoch > current_epoch {
+                return CONTINUE_OP;
+            }
+
+            let subtract_result = self.subtract_payment(
+                &user_address,
+                user_id,
+                &energy_threshold,
+                service_info.payment_type.clone(),
+                fees_contract_address.clone(),
+            );
+            if subtract_result.is_err() {
+                return CONTINUE_OP;
+            }
+
+            let action_results = SC::perform_action(
+                self,
+                user_address.clone(),
+                user_id,
+                &service_info,
+                user_data.borrow(),
+            );
+            if action_results.is_err() {
+                return CONTINUE_OP;
+            }
+
+            // return funds if it didn't work? - discuss
+
+            let interpreted_results = unsafe { action_results.unwrap_unchecked() };
+            if let Some(new_token) = interpreted_results.opt_new_token {
+                self.save_new_token(user_id, new_token);
+            }
+
+            self.send_user_rewards(&user_address, interpreted_results.user_rewards);
+
+            self.last_action_epoch(user_id, service_index)
+                .set(current_epoch);
+
+            CONTINUE_OP
+        });
+
+        if run_result == OperationCompletionStatus::InterruptedBeforeOutOfGas {
+            self.save_progress(&progress);
+        }
+
+        *user_index = progress.additional_data_index;
+
+        run_result
+    }
+
+    fn subtract_payment(
+        &self,
+        user_address: &ManagedAddress,
+        user_id: AddressId,
+        energy_threshold: &BigUint,
+        payment_type: PaymentType<Self::Api>,
+        fees_contract_address: ManagedAddress,
+    ) -> MyVeryOwnScResult<(), ()> {
+        let user_energy = self.get_energy_amount(user_address);
+        let is_premium_user = &user_energy >= energy_threshold;
+        let payment_amount = if is_premium_user {
+            payment_type.amount_for_premium
+        } else {
+            payment_type.amount_for_normal
+        };
+
+        self.call_subtract_payment(
+            fees_contract_address,
+            user_id,
+            payment_type.opt_specific_token,
+            payment_amount,
+        )
+    }
+
+    fn call_subtract_payment(
+        &self,
+        fee_contract_address: ManagedAddress,
+        user_id: AddressId,
+        opt_specific_token: Option<EgldOrEsdtTokenIdentifier>,
+        amount: BigUint,
+    ) -> MyVeryOwnScResult<(), ()> {
+        self.fee_contract_proxy_obj(fee_contract_address)
+            .subtract_payment(user_id, opt_specific_token, amount)
```

Comment:
substract_payment should have input of serviceID only. You do not need to specific token and amount here, as the information of how much tokens and which type of tokens, is in the service information on the subscription contract.

Reply:
Currently not. Each service keeps its own data regarding payment options. If you wish to change that, please let me know.

Reply:
Yes. Information on payment must be strictly in subscription contract. We do not want services to change their prices as they want.